src/Microsoft.OData.Core/ODataReaderCoreAsync.cs

```diff
@@ -175,95 +177,206 @@ protected virtual Task<bool> ReadAtDeltaDeletedLinkImplementationAsync()
+                    this.DecreaseResourceDepth();
+                    result = await this.ReadAtDeletedResourceEndImplementationAsync()
+                        .ConfigureAwait(false);
                     break;
 
                 case ODataReaderState.DeltaLink:
-                    result = this.ReadAtDeltaLinkImplementationAsync();
+                    result = await this.ReadAtDeltaLinkImplementationAsync()
+                        .ConfigureAwait(false);
                     break;
 
                 case ODataReaderState.DeltaDeletedLink:
-                    result = this.ReadAtDeltaDeletedLinkImplementationAsync();
+                    result = await this.ReadAtDeltaDeletedLinkImplementationAsync()
+                        .ConfigureAwait(false);
                     break;
 
                 case ODataReaderState.Exception:    // fall through
                 case ODataReaderState.Completed:
-                    result = TaskUtils.GetFaultedTask<bool>(new ODataException(Strings.ODataReaderCore_NoReadCallsAllowed(this.State)));
-                    break;
+                    throw new ODataException(Strings.ODataReaderCore_NoReadCallsAllowed(this.State));
 
                 default:
                     Debug.Assert(false, "Unsupported reader state " + this.State + " detected.");
-                    result = TaskUtils.GetFaultedTask<bool>(new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataReaderCoreAsync_ReadAsynchronously)));
-                    break;
+                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataReaderCoreAsync_ReadAsynchronously));
             }
 
-            return result.FollowOnSuccessWith(t => t.Result);
+            return result;
+        }
+
+        /// <summary>
+        /// Asynchronously creates a <see cref="Stream"/> for reading a stream property
+        /// when reader in state <see cref="ODataReaderState.Stream"/>.
+        /// </summary>
+        /// <returns>
+        /// A task that represents the asynchronous operation.
+        /// The value of the TResult parameter contains a <see cref="Stream"/> for reading the stream property.
+        /// </returns>
+        protected virtual Task<Stream> CreateReadStreamImplementationAsync()
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Asynchronously creates a <see cref="TextReader"/> for reading a string property
+        /// when reader in state <see cref="ODataReaderState.Stream"/>.
+        /// </summary>
+        /// <returns>
+        /// A task that represents the asynchronous operation.
+        /// The value of the TResult parameter contains a <see cref="TextReader"/> for reading the string property..
+        /// </returns>
+        protected virtual Task<TextReader> CreateTextReaderImplementationAsync()
+        {
+            throw new NotImplementedException();

```

Comment:
Follow up to @xuzhg's [question](https://github.com/OData/odata.net/pull/2335/files#r821216779), will this throw the exception even when you call it synchronously without `await`, is that the intended behaviour?

Reply:
@habbes See my response here https://github.com/OData/odata.net/pull/2335#discussion_r825012910